### PR TITLE
Add `impl_tag!` macro to implement `Tag` for tagged pointer easily

### PR DESCRIPTION
**Pull request by WaffleLapkin**
r? @Nilstrieb 

This should also lifts the need to think about safety from the callers (`impl_tag!` is robust (ish, see the macro issue)) and removes the possibility of making a "weird" `Tag` impl.

### CODE REVIEW

File: compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+/// impl_tag! {
+///     for Flags;
+///     Flags { a: true,  b: true  } <=> 3,
+///     Flags { a: false, b: true  } <=> 2,
+///     Flags { a: true,  b: false } <=> 1,
+///     Flags { a: false, b: false } <=> 0,
+/// }
+/// ```
+///
+// This is supposed to produce a compile error, but does not,
```
Comment (Nilstrieb) [inline compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs:65]: What's `this`, are you missing an example?

File: compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+///
+/// impl_tag! {
+///     for E;
+///     E::A <=> 0,
+/// }
+/// ```
+#[macro_export]
+macro_rules! impl_tag {
+    (
+        for $Self:ty;
```
Comment (Nilstrieb) [inline compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs:114]:
```suggestion
        impl Tag for $Self:ty;
```
nit, makes it nicer to search for impls

**Review by Noratrieb — COMMENTED** [review]
Very cool. I didn't spot any issues except two nits, although I may want to take a second look when you've fixed those.

File: compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs (hunk 1 above)
Comment (WaffleLapkin) [reply compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs:65]: This is talking about the two examples below (note that they are comments, not documentation)

Comment (Nilstrieb) [reply compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs:65]: I think you should remove `Using the same pattern twice results in a compile error:` as long as it's not true, it's a little confusing.

File: compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs (hunk 2 above)
Comment (WaffleLapkin) [reply compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs]: Done

File: compiler/rustc_data_structures/src/tagged_ptr.rs [hunk 3]
```diff
@@ -141,6 +142,40 @@ pub unsafe trait Tag: Copy {
...
+const fn bits_for_tag(mut tag: usize) -> u32 {
+    let mut bits = 0;
+
+    while tag > 0 {
+        bits += 1;
+        tag >>= 1;
+    }
+
+    bits
+}
```
Comment (klensy) [inline compiler/rustc_data_structures/src/tagged_ptr.rs:177]: So, if it can be written shorter (as in doc), why using long version?

Comment (WaffleLapkin) [reply compiler/rustc_data_structures/src/tagged_ptr.rs:177]: hm, for some reason I thought that `leading_zeros` is not `const`, oops

Comment (klensy) [reply compiler/rustc_data_structures/src/tagged_ptr.rs:177]: Someone suggested slightly different version that will be even shorter.
```rust
const fn bits_for_tag(tag: usize) -> u32 {
    if tag == 0 {
        0
    } else {
        (usize::BITS - tag.leading_zeros()) as u32
    }
}
```

Comment (WaffleLapkin) [reply compiler/rustc_data_structures/src/tagged_ptr.rs]: oh, I forgot that `usize::BITS` exists too lmao

Comment (WaffleLapkin) [reply compiler/rustc_data_structures/src/tagged_ptr.rs]: It's actually just `usize::BITS - tag.leading_zeros()`. 

`0.leading_zeros() == usize::BITS`  and both of them are already `u32`.

File: compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+///     impl Tag for E;
+///     E::A <=> 0,
+/// }
+/// ```
+#[macro_export]
+macro_rules! impl_tag {
+    (
+        impl Tag for $Self:ty;
+        $(
+            $($path:ident)::* $( { $( $fields:tt )* })? <=> $tag:literal,
```
Comment (cjgillot) [inline compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs:85]: Is customization according to `$tag` useful?
If not, you can use `${index()}` to generate it automatically, like `rustc_metadata::rmeta::table` does.

Comment (WaffleLapkin) [reply compiler/rustc_data_structures/src/tagged_ptr/impl_tag.rs:85]: I guess the answer is "it could, but probably isn't". I'll work on it next week.

**Pull request merged**